### PR TITLE
Compute polygon orientation using signed area

### LINKED ISSUE #26286: Strange behaviour when trying to index geo_shape polygon over dateline
**Elasticsearch version** (`bin/elasticsearch --version`): 5.5.1

**Plugins installed**: []

**JVM version** (`java -version`): 1.8.0_141

**OS version** (`uname -a` if on a Unix-like system): GNU/Linux

**Description of the problem including expected versus actual behavior**:

When trying to index a geo_shape which crosses the dateline am getting intersection errors. 80% of the issues were solved by subtracting a small amount (1e10^-14) from any value which lies directly on +/-180°. There were no intersections here either but ES seemed to not like the value 180.0.  There are still a few shapes which won't go in. I have investigated to see if there are any intersections and can't find any. Translating the shape to meridian using +180 on -ve values and -180 on +ve values allows it to go in so it is solely a dateline issue. 

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. PUT /test-index
```json
{
   "mappings": {
      "geo_metadata": {
         "_all": {
            "enabled": true
         },
         "properties": {
            "spatial": {
               "properties": {
                  "geometries": {
                     "properties": {
                        "search": {
                           "type": "geo_shape",
                           "tree": "quadtree",
                           "precision": "8km"
                        }
                     }
                  }
               }
            }
         }
      }
   }
}

```
 2. PUT /test-index/geo_metadata/1
```json
{
        "spatial": {
            "geometries": {
                "search":{
                    "coordinates": [[[179.99999999999, -72.661870459646082], [-179.92242898523259, -72.614739463963019], [-179.70509673080812, -72.48077664056396], [-179.49092755066465, -72.346619319446333], [-179.27982991402556, -72.212217663440484], [-179.0801618967584, -72.082942320949627], [-179.91362724448533, -72.077340115037259], [179.99999999999, -72.595885626443078], [179.98777151047784, -72.669300308135334], [179.99999999999, -72.661870459646082]]],
                    "orientation": "counterclockwise",
                    "type": "polygon"
                }
            }
        }
}

```

### PR DESCRIPTION
**Pull request by iverase**
We currently relay in the  southernmost point of a polygon ring to compute the orientation of the polygon. This might lead to wrong results if the point neighbours cross the dateline. This commit changes the computation by using the signed area of the polygon ring.

closes #26286

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java [hunk 1]
```diff
@@ -145,17 +144,27 @@ private static int createEdges(int component, boolean orientation, LinearRing sh
...
+        double minX = Double.POSITIVE_INFINITY;
+        double maxX = Double.NEGATIVE_INFINITY;
+        for (int i = offset; i < offset + length; i++) {
+            signedArea += points[i].getX() * points[i + 1].getY() - points[i].getY() * points[i + 1].getX();
+            minX = Math.min(minX, points[i].getX());
+            maxX = Math.max(maxX, points[i].getX());
+        }
+        if (signedArea == 0) {
+            // Points are collinear or self-intersection
+            throw new InvalidShapeException("Cannot determine orientation: signed area equal to 0");
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/common/geo/GeoPolygonDecomposer.java:157]: Maybe we should add the text of this comment to the exception message?

**Review by imotov — APPROVED** [review]
That's clever. LGTM. It seems like a bug fix, but I wonder if we still should we add this to breaking changes?

**Comment by iverase** [comment]
Why do you think it needs to be added to braking changes? All test pass as they were so it does not seem to be a change in behaviour?

**Comment by imotov** [comment]
I am not totally sure we need a breaking changes note. Can you see a scenario where users would rely on the old behaviour and get a very different results after this change?

**Comment by iverase** [comment]
Theoretically it is possible although I have to been able to find an example. On the other hand, it might be possible that in that case, the behaviour was incorrect but I found sometimes difficult to predict the behaviour in the current implementation.

**Comment by imotov** [comment]
I came to the same conclusion, but thought maybe you have some better examples. I guess, we can treat it as a bug and not put it into breaking changes report then.

**Pull request merged**